lib/si-cli/src/engine/podman_engine.rs

```diff
@@ -23,104 +67,610 @@ impl ContainerEngine for PodmanEngine {
     }
 
     async fn ping(&self) -> CliResult<()> {
-        let ping_info = self.podman.ping().await?;
-        dbg!(&ping_info);
+        self.podman.ping().await?;
         Ok(())
     }
 
     async fn missing_containers(&self) -> Result<Vec<String>, SiCliError> {
-        todo!()
+        let mut missing_containers = Vec::new();
+        let containers = self.downloaded_systeminit_containers_list().await?;
+
+        for name in CONTAINER_NAMES.iter() {
+            let required_container = format!("systeminit/{0}", name);
+            if !containers.iter().any(|c| {
+                c.repo_tags
+                    .iter()
+                    .all(|t| t.contains(&required_container.to_string()))
+            }) {
+                missing_containers.push(required_container.to_string());
+            }
+        }
+
+        Ok(missing_containers)
     }
 
-    async fn download_missing_containers(&self, _missing_containers: Vec<String>) -> CliResult<()> {
-        todo!()
+    async fn download_missing_containers(&self, missing_containers: Vec<String>) -> CliResult<()> {
+        let m = MultiProgress::new();
+        let sty = ProgressStyle::with_template(
+            "{spinner:.red} [{elapsed_precise}] [{wide_bar:.yellow/blue}]",
+        )
+        .unwrap()
+        .progress_chars("#>-");
+
+        let total_size = 100123123;
+
+        println!("Found {0} missing containers", missing_containers.len());
+
+        let mut spawned = Vec::new();
+        for missing_container in missing_containers {
+            let pb = m.add(ProgressBar::new(total_size));
+            pb.set_style(sty.clone());
+
+            let mut message = "Downloading ".to_owned();
+            message.push_str(missing_container.as_str());
+
+            let podman = self.podman.clone();
+
+            let h1 = tokio::spawn(async move {
+                let pull_opts = PullOpts::builder()
+                    // TODO: Can the docker.io/ prefix be omitted?
+                    .reference(format!("docker.io/{}:stable", missing_container))
```

Comment:
I don't believe this TODO blocks the merge. By default Podman expects its image references to be fully-qualified, utilizing a short names configuration file to map short names to fully-qualified names.

```sh
$ head /etc/containers/registries.conf.d/000-shortnames.conf 
[aliases]
  # almalinux
  "almalinux" = "docker.io/library/almalinux"
  "almalinux-minimal" = "docker.io/library/almalinux-minimal"
  # Amazon Linux
  "amazonlinux" = "public.ecr.aws/amazonlinux/amazonlinux"
  # Arch Linux
  "archlinux" = "docker.io/library/archlinux"
  # centos
  "centos" = "quay.io/centos/centos"
```

I'm happy to update this implementation if someone else has a better idea for it.